test/functional/sqlsrv/sqlsrv_ae_datetimes_as_strings.phpt

```diff
@@ -0,0 +1,627 @@
+            if ($errors[0]['SQLSTATE'] != '07006' and
+                $errors[0]['SQLSTATE'] != '22007' and
+                $errors[0]['SQLSTATE'] != '22008' and
+                $errors[0]['SQLSTATE'] != '22018') {
+                print_r($errors);
+                fatalError("Unexpected error");
+            }
+        }
+    }
+}
+
+// Compare dates retrieved from the database with the date used for testing.
+// $expectedDateTime is an array of strings like '2002-01-31 23:59:59.049876'
+// $retrievedDateTime is date/time string of format either 'Y-m-d H:i:s.u' or
+// 'Y-m-d H:i:s.u P', which is the format returned when $returnDatesAsStrings 
+// is true, or by the date_format() calls in FetchDatesAndOrTimes when a PHP 
+// DateTime object is retrieved (i.e. when $returnDatesAsStrings is false),
+// unless $datetimetype is date or time. In those cases:
+// If $datetimetype is date and $returnDatesAsStrings is false:
+//   The date is as expected, the time defaults to 00:00:00.0000
+// If $datetimetype is time and $returnStrings is false:
+//   The date defaults to the current date, the time is as expected. 
+// If $datetimetype is date and $returnStrings is true:
+//   $retrievedDateTime is only a date.
+// If $datetimetype is 'time' and $returnStrings is true:
+//   $retrievedDateTime is only a time.
+function CompareDateTime($datetimetype, $returnStrings, &$expectedDateTime, $retrievedDateTime)
+{
+    $expected_date_time = array();
+
+    // Split each element of the testing date/time into
+    // [0]:date, [1]:time, and possibly [2]:timezone offset 
+    for ($i = 0; $i < sizeof($expectedDateTime); ++$i) {
+        $expected_date_time[] = explode(" ", $expectedDateTime[$i]);
+    }
+    
+    // If $retrievedDateTime is a string of format 'Y-m-d H:i:s.u' or 'Y-m-d H:i:s.u P',
+    // split it into [0]:date, [1]:time, and possibly [2]:timezone offset
+    $retrieved_date_time = explode(" ", $retrievedDateTime);
+   
+    if ($returnStrings == true) {
+        switch ($datetimetype) {
+            case 'date':
+                // Direct comparison of retrieved date and expected date
+                if ($retrievedDateTime != $expected_date_time[0][0]) {
+                    fatalError("Dates do not match!");
+                }                    
+                break;
+            case 'time': 
+                // Compare SQL time with expected time. The expected time was input
+                // with an accuracy of microseconds and the SQL Servertime type has
+                // accuracy to 100 ns, so times are returned with an extra zero. For
+                // comparison the zero is appended to the times in expected_time_date.
+                if ($retrievedDateTime != $expected_date_time[0][1]."0" and
+                    $retrievedDateTime != $expected_date_time[1][1]."0" and
+                    $retrievedDateTime != $expected_date_time[2][1]."0" and
+                    $retrievedDateTime != $expected_date_time[3][1]."0" ) {
```

Comment:
can remove this lost condition since `$expected_date_time[3][1]` corresponds to format `$year."-".$month."-".$day." ".$hour.":".$minute.":00.000000"` and that should never be received with time type.

Reply:
Seconds are optional with the time type, so it's possible that time fields do not have seconds.